feature_examples/tensorflow2/pipelining/pipelining.py

```diff
@@ -0,0 +1,425 @@
+"""
+Copyright (c) 2021 Graphcore Ltd. All rights reserved.
+"""
+"""
+# TensorFlow 2: Model Parallelism with IPU Pipelining
+
+In this tutorial you will train a selection of simple fully connected models
+on the MNIST numeral data set and see how training can be parallelised over
+multiple IPU devices.
+"""
+"""
+## Upgrading to TensorFlow 2
+
+Considering that IPU computation can be enabled on both TensorFlow 1 
+and Tensorflow 2 it is necessary to explain the major differences between them
+and how it affects implementation of IPU specific code.
+
+### Device scopes
+In IPU APIs for TF2, the scope context `ipu.scopes.ipu_scope(device_id)` was
+replaced with a strategy context `ipu.ipu_strategy.IPUStrategy().scope().
+
+### Training loop
+Since TF2 moved in the direction of eager execution, we no longer are required
+to create sessions and use them as context (`with tf.Session()...`). Instead, 
+when using the Keras API, we can use the model instance directly and invoke
+`model.compile()`, `model.fit()`, and `model.predict()` methods without
+specifing explicitly the training loop. To enable IPUs, it is just required
+that these invocations are executed under `IPUStrategy` scope.
+
+### Keras extensions to facilitate IPU computing
+You can find the main documentation on the [GraphCore Keras for IPU](https://docs.graphcore.ai/projects/tensorflow-user-guide/en/latest/keras_tf2.html) page.
+The framework has been extended to enable IPU devices usage and configuration.
+All the new code can be found within the `tensorflow.python.ipu` package.
+
+### TF2 specific changes
+There is a guide prepared by the TensorFlow team to conduct migration between
+versions of TensorFlow library, which you can study [here](https://www.tensorflow.org/guide/migrate).
+
+A very exhaustive comparison of both versions can be found [here](https://www.tensorflow.org/guide/migrate/tf1_vs_tf2).
+"""
+
+"""
+## Pipelining features
+In this tutorial, we will create models using the Keras Model class and IPU 
+pipelining features.
+
+We are going to use Pipeline Stages to assign operations to devices and to
```

Comment:
```suggestion
pipelining features. We are going to use Pipeline Stages to assign operations to devices and to
```
I think this will look better in one line